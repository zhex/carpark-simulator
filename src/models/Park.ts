import { IPosition } from './interfaces/IPosition';

export class Park {
    constructor(public rows: number, public cols: number) {}

    public inZone(pos: IPosition): boolean {
        return pos.x >= 0 && pos.x < this.rows && pos.y >= 0 && pos.y < this.cols;
    }
}
