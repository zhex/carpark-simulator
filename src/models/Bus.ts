import { IPosition } from './interfaces/IPosition';
import { Park } from './Park';

export enum Direction {
    North = 1,
    East,
    South,
    West,
}

const directionText = ['NORTH', 'EAST', 'SOUTH', 'WEST'];

export class Bus {
    public position: IPosition;
    public faceTo: Direction;

    constructor(public park: Park) {}

    public place(
        x: number | string,
        y: number | string,
        faceTo: Direction | string,
    ): Bus {
        this.position = {
            x: typeof x === 'string' ? parseInt(x, 10) : x,
            y: typeof y === 'string' ? parseInt(y, 10) : y,
        };

        if (typeof faceTo === 'string') {
            const idx = directionText.indexOf(faceTo);
            this.faceTo = idx + 1;
        } else {
            this.faceTo = faceTo;
        }

        return this;
    }

    public move(): Bus {
        const next = { ...this.position };

        switch (this.faceTo) {
            case Direction.North:
                next.y++;
                break;
            case Direction.East:
                next.x++;
                break;
            case Direction.South:
                next.y--;
                break;
            case Direction.West:
                next.x--;
                break;
        }
        if (this.park.inZone(next)) {
            this.position = next;
        }
        return this;
    }

    public left(): Bus {
        this.faceTo = this.faceTo - 1;
        if (this.faceTo < Direction.North) {
            this.faceTo = Direction.West;
        }
        return this;
    }

    public right(): Bus {
        this.faceTo = this.faceTo + 1;
        if (this.faceTo > Direction.West) {
            this.faceTo = Direction.North;
        }
        return this;
    }

    public report(): Bus {
        const msg = [
            this.position.x,
            this.position.y,
            directionText[this.faceTo - 1],
        ].join(',');
        console.log(msg);
        return this;
    }

    public command(cmd: string, args: any[] = []): Bus {
        (this as any)[cmd.toLowerCase()].apply(this, args);
        return this;
    }

    public clone(): Bus {
        const b = new Bus(this.park);
        b.position = { ...this.position };
        b.faceTo = this.faceTo;
        return b;
    }
}
