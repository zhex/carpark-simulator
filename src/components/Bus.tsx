import { Direction } from 'models/Bus';
import * as React from 'react';
import styled from 'styled-components';

export interface IBusProps {
    x: number;
    y: number;
    faceTo: Direction;
    className?: string;
    cellSize: number;
    width?: number;
    height?: number;
}

const Bus: React.SFC<IBusProps> = props => <div className={props.className} />;

const StyledBus = styled(Bus)`
    position: absolute;
    float: left;
    border-style: solid;
    border-width: 0 ${p => `${p.width! / 2}px`} ${p => `${p.height}px`};
    border-color: transparent transparent #dd4397 transparent;
    left: ${p => `${(p.cellSize - 20) / 2}px`};
    bottom: ${p => `${(p.cellSize - 15 / 2) / 2}px`};
    transform: translate(
            ${p => `${p.x * p.cellSize}px`},
            ${p => `${-1 * p.y * p.cellSize}px`}
        )
        rotate(${p => ((p.faceTo || 1) - 1) * 90}deg);
    transition: 0.3s all ease;
`;

StyledBus.defaultProps = {
    width: 20,
    height: 15,
};

export default StyledBus;
