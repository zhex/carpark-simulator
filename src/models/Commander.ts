import { readFileSync } from 'fs';
import { Bus } from './Bus';
import { ICmd } from './interfaces/ICmd';

export class Commander {
    public static load(content: string): Commander {
        const commander = new Commander();
        commander.commands = content
            .split('\n')
            .filter(c => c !== '')
            .map(c => commander.parse(c));
        return commander;
    }

    public static loadFile(filePath: string): Commander {
        const content = readFileSync(filePath).toString();
        return Commander.load(content);
    }

    public commands: ICmd[];

    public execute(bus: Bus): void {
        this.commands.forEach(cmd => bus.command(cmd.name, cmd.args));
    }

    public parse(cmdStr: string): ICmd {
        const tmp = cmdStr.split(' ');
        return {
            name: tmp[0],
            args: tmp[1] ? tmp[1].split(',') : [],
        };
    }
}
