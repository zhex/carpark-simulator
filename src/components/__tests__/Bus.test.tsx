import 'jest-styled-components';
import * as React from 'react';
import { create } from 'react-test-renderer';
import { Direction } from '../../models/Bus';

import Bus from '../Bus';

it('Bus Component', () => {
    const tree = create(<Bus x={5} y={5} cellSize={50} faceTo={Direction.North} />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('left', '15px');
    expect(tree).toHaveStyleRule('bottom', '21.25px');
    expect(tree).toHaveStyleRule('transform', 'translate( 250px, -250px ) rotate(0deg)');
});
