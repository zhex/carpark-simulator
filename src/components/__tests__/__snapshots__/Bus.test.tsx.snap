// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bus Component 1`] = `
.c0 {
  position: absolute;
  float: left;
  border-style: solid;
  border-width: 0 10px 15px;
  border-color: transparent transparent #dd4397 transparent;
  left: 15px;
  bottom: 21.25px;
  -webkit-transform: translate( 250px, -250px ) rotate(0deg);
  -ms-transform: translate( 250px, -250px ) rotate(0deg);
  transform: translate( 250px, -250px ) rotate(0deg);
  -webkit-transition: 0.3s all ease;
  transition: 0.3s all ease;
}

<div
  className="c0"
/>
`;
